### PR TITLE
Common development raw

### PR DESCRIPTION
**Pull request by igasm**
Classes to conversion between JSONString and game objects. This utils are needed for client-server communication.

### CODE REVIEW

File: common/pom.xml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+                    <source>1.8</source>
+                    <target>1.8</target>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+    <packaging>jar</packaging>
+    <name>common</name>
+
+    <dependencies>
```
Comment (szczepanskikrs) [inline common/pom.xml]: No properties, define properties constants in parent POM

File: common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import com.fasterxml.jackson.databind.SerializationFeature;
+
+import java.util.Optional;
+
+class JsonMarshaller {
+
+    Optional<String> writeToJsonString(Messagable messegable){
+        ObjectMapper objectMapper = new ObjectMapper().configure(SerializationFeature.WRAP_ROOT_VALUE, true);
+
+        String jsonString = null;
```
Comment (szczepanskikrs) [inline common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java]: Is null good idea?

File: common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+package battleships.communication.jsonHandlers;
+
+import battleships.communication.Messagable;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+
+import java.util.Optional;
+
```
Comment (szczepanskikrs) [inline common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java]: JavaDocs? Now or later?

File: common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+import java.io.IOException;
+import java.util.*;
+
+class JsonUnmarshaller {
+
+    private final Map<String, Class<? extends Messagable>> registry;
+
+    public JsonUnmarshaller(Map<String, Class<? extends Messagable>> registry) {
```
Comment (szczepanskikrs) [inline common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: Change string to enum?

File: common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        this.registry = registry;
+    }
+
+    List<Messagable> readFromJSONString(String message) throws IOException, ClassNotFoundException {
+        ObjectMapper objectMapper = new ObjectMapper().configure(DeserializationFeature.UNWRAP_ROOT_VALUE, true);
+        List<String> rootClassNames  = getRootClassName(message);
+        List<Messagable> messegableList = new LinkedList<Messagable>();
+
+        for(String rootClassName : rootClassNames){
```
Comment (szczepanskikrs) [inline common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: Maybe replace with lambda?

File: common/src/main/java/battleships/communication/messages/GoodByeMessage.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+    public GoodByeMessage(String body){
+        this.body = body;
+    }
+
+    public String getBody() {
+        return body;
+    }
+
+    public void setBody(String body) {
```
Comment (szczepanskikrs) [inline common/src/main/java/battleships/communication/messages/GoodByeMessage.java]: Do we need setters or can we only use constructor?

File: common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+class JsonMarshaller {
+
+    Optional<String> writeToJsonString(Messagable messegable){
```
Comment (robwrobel) [inline common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java]: convertToJsonString

File: common/src/main/java/battleships/communication/messages/WelcomeMessage.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+    public WelcomeMessage(String body){
+        this.body = body;
+    }
+
+    public String getBody() {
+        return body;
+    }
+
+    public void setBody(String body) {
```
Comment (szczepanskikrs) [inline common/src/main/java/battleships/communication/messages/WelcomeMessage.java]: Do we need setters or can we use only constructor?

File: common/src/test/java/battleships/communication/jsonHandlers/JsonMarshallerTest.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+                {new GoodByeMessage(), "{\"GoodByeMessage\":{\"body\":\"\"}}"}
+        };
+    }
+
+    @Test(dataProvider = "messagesPool")
+    public void givenObjectOfMessagable_jsonBodyShouldContainsClassNameAndItsFields(Messagable messagable, String expectedJsonString){
+        JsonMarshaller jsonMarshaller = new JsonMarshaller();
+        Optional<String> opt =  jsonMarshaller.writeToJsonString(messagable);
+        String actualJsonString = opt.get();
+        assertThat(actualJsonString).isEqualTo(expectedJsonString);
```
Comment (szczepanskikrs) [inline common/src/test/java/battleships/communication/jsonHandlers/JsonMarshallerTest.java]: Nice :+1:  I like that you changed tests to match the standard :)

File: common/src/test/java/battleships/communication/jsonHandlers/JsonMarshallerTest.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+                {new WelcomeMessage(), "{\"WelcomeMessage\":{\"body\":\"\"}}"},
+                {new GoodByeMessage("good bye!"), "{\"GoodByeMessage\":{\"body\":\"good bye!\"}}"},
+                {new GoodByeMessage(""), "{\"GoodByeMessage\":{\"body\":\"\"}}"},
+                {new GoodByeMessage(), "{\"GoodByeMessage\":{\"body\":\"\"}}"}
+        };
+    }
+
+    @Test(dataProvider = "messagesPool")
+    public void givenObjectOfMessagable_jsonBodyShouldContainsClassNameAndItsFields(Messagable messagable, String expectedJsonString){
+        JsonMarshaller jsonMarshaller = new JsonMarshaller();
```
Comment (szczepanskikrs) [inline common/src/test/java/battleships/communication/jsonHandlers/JsonMarshallerTest.java]: Put creation of objects this in @BeforeTest ?

File: common/src/test/java/battleships/communication/jsonHandlers/JsonUnmarshallerTest.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.assertj.core.api.AssertionsForClassTypes.assertThat;
+
+public class JsonUnmarshallerTest {
+
+    private Map<String, Class<? extends Messagable>> registry = new HashMap<>();
```
Comment (szczepanskikrs) [inline common/src/test/java/battleships/communication/jsonHandlers/JsonUnmarshallerTest.java]: Enum?

File: common/src/test/java/battleships/communication/jsonHandlers/JsonUnmarshallerTest.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    protected void beforeTest(){
+        registry.put("WelcomeMessage", WelcomeMessage.class);
+        registry.put("GoodByeMessage", GoodByeMessage.class);
+        jsonUnmarshaller = new JsonUnmarshaller(registry);
+    }
+
+    @DataProvider
+    public Object[][] jsonBodyAndRootClassesNamesPoll(){
+        return new Object[][] {
+                {"{\"WelcomeMessage\":{\"body\":\"hello\"}}", new String[]{"WelcomeMessage"}},
```
Comment (szczepanskikrs) [inline common/src/test/java/battleships/communication/jsonHandlers/JsonUnmarshallerTest.java]: Is there any way to test this without hard coding this values for every message?

File: common/pom.xml [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>1.8</source>
+                    <target>1.8</target>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+    <packaging>jar</packaging>
```
Comment (szczepanskikrs) [inline common/pom.xml]: Please place it in the top of pom ?

File: common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+                messegableList.add(objectMapper.readValue(message, registry.get(rootClassName)));
+        }
+        return messegableList;
+    }
+
+    List<String> getRootClassName(String message) throws IOException {
+        List<String> rootClassesNames = new LinkedList<String>();
```
Comment (szczepanskikrs) [inline common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: Why linked list?

File: common/pom.xml [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+    <name>common</name>
```
Comment (szczepanskikrs) [inline common/pom.xml]: Move up

File: common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+class JsonUnmarshaller {
+
+    private final Map<String, Class<? extends Messagable>> registry;
+
+    public JsonUnmarshaller(Map<String, Class<? extends Messagable>> registry) {
+        this.registry = registry;
+    }
+
+    List<Messagable> readFromJSONString(String message) throws IOException, ClassNotFoundException {
```
Comment (robwrobel) [inline common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: convertToMessagableList

File: common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+                messegableList.add(objectMapper.readValue(message, registry.get(rootClassName)));
+        }
+        return messegableList;
+    }
+
+    List<String> getRootClassName(String message) throws IOException {
+        List<String> rootClassesNames = new LinkedList<String>();
+
+        ObjectMapper objectMapper = new ObjectMapper();
+        ObjectNode root = (ObjectNode) objectMapper.readTree(message);
```
Comment (szczepanskikrs) [inline common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: Misleading name root

File: common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java (hunk 14 above)
Comment (szczepanskikrs) [inline common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: Misleading name?

File: common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java (hunk 7 above)
Comment (igasm) [reply common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java]: done

File: common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java (hunk 16 above)
Comment (igasm) [reply common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: done

File: common/pom.xml (hunk 13 above)
Comment (igasm) [reply common/pom.xml]: done

File: common/pom.xml (hunk 15 above)
Comment (igasm) [reply common/pom.xml]: done

File: common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java (hunk 3 above)
Comment (igasm) [reply common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java]: later

File: common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java (hunk 2 above)
Comment (igasm) [reply common/src/main/java/battleships/communication/jsonHandlers/JsonMarshaller.java]: changed to empty string

File: common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java (hunk 14 above)
Comment (igasm) [reply common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: changed to arrayList

Comment (igasm) [reply common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: changed

File: common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java (hunk 17 above)
Comment (igasm) [reply common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: changed

File: common/src/main/java/battleships/communication/messages/GoodByeMessage.java (hunk 6 above)
Comment (igasm) [reply common/src/main/java/battleships/communication/messages/GoodByeMessage.java]: getters/setters are needed by Jackson lib

File: common/src/main/java/battleships/communication/messages/WelcomeMessage.java (hunk 8 above)
Comment (igasm) [reply common/src/main/java/battleships/communication/messages/WelcomeMessage.java]: getters/setters are needed by Jackson lib

File: common/src/test/java/battleships/communication/jsonHandlers/JsonMarshallerTest.java (hunk 10 above)
Comment (igasm) [reply common/src/test/java/battleships/communication/jsonHandlers/JsonMarshallerTest.java]: done

File: common/src/test/java/battleships/communication/jsonHandlers/JsonUnmarshallerTest.java (hunk 11 above)
Comment (igasm) [reply common/src/test/java/battleships/communication/jsonHandlers/JsonUnmarshallerTest.java]: maybe later

File: common/src/test/java/battleships/communication/jsonHandlers/JsonUnmarshallerTest.java (hunk 12 above)
Comment (igasm) [reply common/src/test/java/battleships/communication/jsonHandlers/JsonUnmarshallerTest.java]: At that point, I this that hard coding is ok.

**Comment by igasm** [comment]
I added most changes as Robert and Krzysztof requested, more in comments.

File: common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java (hunk 4 above)
Comment (igasm) [reply common/src/main/java/battleships/communication/jsonHandlers/JsonUnmarshaller.java]: i thinks there are no need at that point

**Pull request merged**